// @medusajs/medusa/dist/api/middlewares/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformQuery = exports.transformBody = exports.transformIncludesOptions = exports.doesConditionBelongToDiscount = exports.getRequestedBatchJob = exports.canAccessBatchJob = void 0;
var authenticate_1 = __importDefault(require("./authenticate"));
var authenticate_customer_1 = __importDefault(require("./authenticate-customer"));
var await_middleware_1 = __importDefault(require("./await-middleware"));
var normalized_query_1 = __importDefault(require("./normalized-query"));
var require_customer_authentication_1 = __importDefault(require("./require-customer-authentication"));
var can_access_batch_job_1 = require("./batch-job/can-access-batch-job");
Object.defineProperty(exports, "canAccessBatchJob", { enumerable: true, get: function () { return can_access_batch_job_1.canAccessBatchJob; } });
var get_requested_batch_job_1 = require("./batch-job/get-requested-batch-job");
Object.defineProperty(exports, "getRequestedBatchJob", { enumerable: true, get: function () { return get_requested_batch_job_1.getRequestedBatchJob; } });
var does_condition_belong_to_discount_1 = require("./discount/does-condition-belong-to-discount");
Object.defineProperty(exports, "doesConditionBelongToDiscount", { enumerable: true, get: function () { return does_condition_belong_to_discount_1.doesConditionBelongToDiscount; } });
var transform_includes_options_1 = require("./transform-includes-options");
Object.defineProperty(exports, "transformIncludesOptions", { enumerable: true, get: function () { return transform_includes_options_1.transformIncludesOptions; } });
var transform_body_1 = require("./transform-body");
Object.defineProperty(exports, "transformBody", { enumerable: true, get: function () { return transform_body_1.transformBody; } });
var transform_query_1 = require("./transform-query");
Object.defineProperty(exports, "transformQuery", { enumerable: true, get: function () { return transform_query_1.transformQuery; } });
exports.default = {
    authenticate: authenticate_1.default,
    authenticateCustomer: authenticate_customer_1.default,
    requireCustomerAuthentication: require_customer_authentication_1.default,
    normalizeQuery: normalized_query_1.default,
    wrap: await_middleware_1.default,
};
